artiq/firmware/libboard_misoc/flash.rs

```diff
@@ -0,0 +1,159 @@
+extern crate crc;
+use core::fmt;
+
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Error {
+    AlreadyLocked,
+    NoFlash,
+    WrongPartition,
+    WriteFail { sector: usize },
+    CorruptedFirmware,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Error::AlreadyLocked =>
+                write!(f, "attempt at reentrant access"),
+            &Error::NoFlash =>
+                write!(f, "flash memory is not present"),
+            &Error::WrongPartition =>
+                write!(f, "Unknown partition"),
+            &Error::WriteFail { sector } =>
+                write!(f, "Flash writing failed in sector {}", sector),
+            &Error::CorruptedFirmware =>
+                write!(f, "Corrupted file"),
+        }
+    }
+}
+
+#[cfg(has_spiflash)]
+mod imp {
+    use core::str;
+    use cache;
+    use spiflash;
+    use super::Error;
+
+    use core::slice;
+    use core::cmp;
+    use csr;
+    use byteorder::{ByteOrder, BigEndian};
+    use flash::crc::crc32;
+
+    const SIZE: usize = spiflash::SECTOR_SIZE;
+    const ADDR_GATEWARE: usize = 0x0;
+    const ADDR_BOOTLOADER: usize = 0x400000;
+    const ADDR_FIRMWARE: usize = ::mem::FLASH_BOOT_ADDRESS;
+
+    mod lock {
```

Comment:
Unnecessary AFAICT and in any case would not belong here.

Reply:
Do you mean the mod lock? It is from the config.rs, but I modified the functions so I cannot use that one.

Reply:
Ah, I see, it's because of the weird thread implementation we have.
I still don't see why it really has to be modified and it would be better to share it with config.rs.

Reply:
In config.rs, the address is fixed to storage. But in my case, the address could be one of gateware/bootloader/firmware. So I added the address as a parameter. I can change this part in config.rs and share it.